### ISSUE TITLE
Parity with containers

### ISSUE THREAD

**Issue by hseg** [haskell-unordered-containers/unordered-containers#302]
> Hm, consistency with `containers` is also nice. `containers` is huge though, and that also comes with costs with regard to discoverability and maintainability IMHO.
> 
> _Originally posted by @sjakobi in https://github.com/haskell-unordered-containers/unordered-containers/issues/245#issuecomment-637489756_

That said, I've just diffed the APIs, and am noting it here for posterity:

- [ ] `adjustWithKey`
- [ ] `assocs`
- [ ] `differenceWithKey`
- [ ] `disjoint`
- [ ] `insertLookupWithKey`
- [ ] `insertWithKey`
- [ ] `isProperSubmapOf`
- [ ] `isProperSubmapOfBy`
- [ ] `mapAccumRWithKey`
- [ ] `mapAccumWithKey`
- [ ] `mapEitherWithKey`
- [ ] `mapKeysMonotonic`
- [ ] `mapKeysWith`
- [ ] `mergeWithKey`
- [ ] `partitionWithKey`
- [ ] `restrictKeys`
- [ ] `traverseMaybeWithKey`
- [ ] `updateLookupWithKey`
- [ ] `updateWithKey`
- [ ] `withoutKeys`

The others are either a) low-level indexing or b) require an ordering on keys, so I think these are the only ones that might reasonably be cherrypicked in. Note that we already have `WithKey` functions, so that can't be what's blocking these.
One possible counterargument is that some of these don't cross the Fairbairn threshold. In that case, that's grounds for opening tickets against `containers` to remove them. In any case, this discussion needs to start somewhere, and I'm starting it here.

**Comment by sjakobi** [comment]
Thanks for the initiative and the list, @hseg! :)

I assume you've compared the `HashMap` API against `Map` so far? A similar list for `HashSet` would be nice to have, too.

I think most of these would be good additions – with the following exceptions:

* `mapAccum*`: So far I haven't understood how these functions could be useful when the iteration order is unspecified. See https://github.com/haskell-unordered-containers/unordered-containers/pull/91#issuecomment-638779035.
* `notMember`: I think users can simply use `member` and negate the result. IMHO an API addition wouldn't cross the Fairbairn threshold in this case.

**Comment by hseg** [comment]
Fair, missed #89 in searching. Agree about `notMember`, the more I look into this and the more inconsistently bloated the `containers` API seems. Re `mapAccum`, I miscategorized it as being independent of order.
Though that does raise the question of how you'd implement it in terms of `traverseWithKey`, assuming a combinator commutative in the values.

Re drop-in compatibility with `containers`, it'd be nice to convert `IsMap` from `mono-traversable` into a backpack signature to avoid instantiation costs. And then we could have `container-utils` that exports all those specializations that are of questionable use. Something to shoot for?

**Comment by sjakobi** [comment]
> Re drop-in compatibility with `containers`, it'd be nice to convert `IsMap` from `mono-traversable` into a backpack signature to avoid instantiation costs. And then we could have `container-utils` that exports all those specializations that are of questionable use. Something to shoot for?

Yes, that sounds desirable, and, to some extent, it already exists: https://github.com/kowainik/containers-backpack. I've never taken a very close look at it though.

**Comment by treeowl** [comment]
Ed Kmett also has unpacked-containers or something.

**Comment by hseg** [comment]
`containers-backpack` is proof-of-concept level atm, but the idea should work.
In fact, it would generally be nice to be able to move between class/instance abstractions and signature/modules abstractions. Perhaps a summer project.

**Comment by layus** [comment]
Just chiming in with my own concern here, but I am missing `traverseMaybeWithKey` because it cannot be reimplemented efficiently in terms of other primitives. I need to `fmap filter . traverse . map`. Well traverse subsumes map, so it is rather `fmap filter . traverse`. But the composition builds an unnecessary intermediate hashmap which `traverseMaybeWithKey` could avoid.

**Comment by treeowl** [comment]
Yes, that's a good one to add.
